src/controllers/portfolio/portfolio.ts

```diff
@@ -440,178 +526,138 @@ export class PortfolioController extends EventEmitter {
-                      accountOps: currentAccountOps
-                    }
-                  }),
-                  isEOA: !isSmartAccount(selectedAccount),
-                  additionalHints
-                },
-                forceUpdate
-              )
-            : Promise.resolve(false)
-        ])
-
-        // Persist latest state in previousHints in the disk storage for further requests
-        if (
-          isSuccessfulLatestUpdate &&
-          !areAccountOpsChanged &&
-          accountState[network.id]?.result?.hintsFromExternalAPI
-        ) {
-          const updatedStoragePreviousHints = getUpdatedHints(
-            accountState[network.id]!.result!.hintsFromExternalAPI as ExternalHintsAPIResponse,
-            accountState[network.id]!.result!.tokens,
-            network.id,
-            this.#previousHints,
-            key,
-            this.tokenPreferences
-          )
+              ? this.updatePortfolioState(
+                  accountId,
+                  pendingState,
+                  network,
+                  portfolioLib,
+                  {
+                    blockTag: 'pending',
+                    previousHints: fallbackHints,
+                    ...(currentAccountOps && {
+                      simulation: {
+                        account: selectedAccount,
+                        accountOps: currentAccountOps
+                      }
+                    }),
+                    isEOA: !isSmartAccount(selectedAccount),
+                    additionalHints
+                  },
+                  forceUpdate
+                )
+              : Promise.resolve(false)
+          ])
+
+          // Persist latest state in previousHints in the disk storage for further requests
+          if (
+            isSuccessfulLatestUpdate &&
+            !areAccountOpsChanged &&
+            accountState[network.id]?.result?.hintsFromExternalAPI
+          ) {
+            const updatedStoragePreviousHints = getUpdatedHints(
+              accountState[network.id]!.result!.hintsFromExternalAPI as ExternalHintsAPIResponse,
+              accountState[network.id]!.result!.tokens,
+              network.id,
+              this.#previousHints,
+              key,
+              this.tokenPreferences
+            )
+
+            this.#previousHints = updatedStoragePreviousHints
+            await this.#storage.set('previousHints', updatedStoragePreviousHints)
+          }
 
-          this.#previousHints = updatedStoragePreviousHints
-          await this.#storage.set('previousHints', updatedStoragePreviousHints)
+          // We cache the previously simulated AccountOps
+          // in order to compare them with the newly passed AccountOps before executing a new updatePortfolioState.
+          // This allows us to identify any differences between the two.
+          if (currentAccountOps) {
+            pendingState[network.id]!.accountOps = currentAccountOps
+          }
         }
```

Comment:
Perhaps this can be moved to a method as well?

Reply:
I considered the options and prefer to pull out only updatePortfolioState as a standalone method. Otherwise, it might be difficult to follow the implementation logic. Having updateSelectedAccount as the main method with an empty inner body that just calls other functions makes it harder to understand the order of execution. If the method becomes significantly larger, we will definitely refactor it accordingly.